### PR TITLE
Automated test tutorial

### PR DESCRIPTION
**Pull request by lpil**
Hello! I wrote a simple tutorial on testing as part of #98 Let me know your thoughts. :)

Cheers,
Louis

### CODE REVIEW

**Comment by agrafix** [comment]
Cool, thanks! I will look at this sometime this week, but I'm really busy atm.

**Comment by lpil** [comment]
No problem, thanks :)

File: docs/tutorials/testing.md [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+    main: Spec.hs
+    source-dirs: test
+    dependencies:
+      - hspec       # Test deps here!
+      - hspec-wai   # Test deps here!
+```
+
+## Creating tests
+
+Now write a simple test to `test/Spec.hs` to ensure that the Hspec test
```
Comment (agrafix) [inline docs/tutorials/testing.md]: "Now we create a file `test/Spec.hs`, start with the test setup and add a basic test"

File: docs/tutorials/testing.md [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+           text "Hello World!"
+       get ("hello" <//> var) $ \name ->
+           do text ("Hello " <> name)
+```
+
+`app` has been renamed to `routes`, and the `main` function has been split
+into two. The new `main` is responsible for only the running of the
+application, with the new `app` function being responsible for configuring the
+application. Through the new `app` we can access an `IO Middleware` which we
+can convert into an `IO Application` in the tests.
```
Comment (agrafix) [inline docs/tutorials/testing.md]: ... using `spockAsApp` ...

File: docs/tutorials/testing.md [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+spec =
+    with (spockAsApp app) $
+        do describe "GET /" $
+               do it "serves the home page" $
+                      get "/" `shouldRespondWith` "Hello World" {matchStatus = 200}
+           describe "GET /hello/:name" $
+               do it "returns hello to spock" $
+                      get "/hello/spock" `shouldRespondWith` "Hello spock"
+                  it "returns hello to uhura" $
+                      get "/hello/spock" `shouldRespondWith` "Hello uhura"
```
Comment (agrafix) [inline docs/tutorials/testing.md]: This test will fail?

Comment (agrafix) [inline docs/tutorials/testing.md]: Maybe explain the testing combinators a bit?

File: docs/tutorials/testing.md [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+world!", and `/hello/:name`, which responds with "Hello `$NAME`", where name
+is the value of the second path segment.
+
+It type checks, so we are confident that the application will run without
+crashing, but we might want some tests to check that the business logic has
+been correctly implemented. We can do this with a little help from the
+[Hspec][hspec] and [Hspec-Wai][hspec-wai] libraries.
+
+Add Hspec and Hspec-Wai to your tests dependencies in your project's
+`package.yaml` file or Cabal file. If you're using `package.yaml` it may look
```
Comment (lpil) [inline docs/tutorials/testing.md]: I'm afraid I'm new to Haskell and am not very familiar with Cabalfiles, how does this look?

**Comment by lpil** [comment]
Great feedback, thank you. Let me know how this new version looks.

**Comment by agrafix** [comment]
Thanks! I'm merging this now. If you have any improvements, feel free to send a PR :-)

**Comment by lpil** [comment]
Fab, thank you!

**Pull request merged**